File: box_pricer.py
def calculate_box_sell_price(quotes, k1, k2):
    """
    Box Sell formula:

    -Call(K1) + Put(K1) + Call(K2) - Put(K2)

    Execution logic:

    -Call(K1)  means sell Call(K1) -> use bid
    +Put(K1)   means buy Put(K1)   -> use ask
    +Call(K2)  means buy Call(K2)  -> use ask
    -Put(K2)   means sell Put(K2)  -> use bid
    """

    c1 = quotes[(k1, "CE")]
    p1 = quotes[(k1, "PE")]
    c2 = quotes[(k2, "CE")]
    p2 = quotes[(k2, "PE")]

    zero_legs = []
    if not c1["bid"]: zero_legs.append(f"CE({k1}) bid=0")
    if not p1["ask"]: zero_legs.append(f"PE({k1}) ask=0")
    if not c2["ask"]: zero_legs.append(f"CE({k2}) ask=0")
    if not p2["bid"]: zero_legs.append(f"PE({k2}) bid=0")
    if zero_legs:
        raise ValueError(f"Zero price leg(s) in BOX SELL ({k1}/{k2}): {', '.join(zero_legs)}")

    box_sell_price = (
        c1["bid"]
        - p1["ask"]
        - c2["ask"]
        + p2["bid"]
    )

    legs = {
        "sell_call_k1": {
            "contract": c1["display_name"],
            "price_used": c1["bid"]
        },
        "buy_put_k1": {
            "contract": p1["display_name"],
            "price_used": p1["ask"]
        },
        "buy_call_k2": {
            "contract": c2["display_name"],
            "price_used": c2["ask"]
        },
        "sell_put_k2": {
            "contract": p2["display_name"],
            "price_used": p2["bid"]
        }
    }

    return box_sell_price, legs

def calculate_box_long_price(quotes, k1, k2):
    """
    Box Long formula:

    +Call(K1) - Put(K1) - Call(K2) + Put(K2)

    Executable price:

    B_buy = C(K1)_ask - P(K1)_bid - C(K2)_bid + P(K2)_ask
    """

    c1 = quotes[(k1, "CE")]
    p1 = quotes[(k1, "PE")]
    c2 = quotes[(k2, "CE")]
    p2 = quotes[(k2, "PE")]

    zero_legs = []
    if not c1["ask"]: zero_legs.append(f"CE({k1}) ask=0")
    if not p1["bid"]: zero_legs.append(f"PE({k1}) bid=0")
    if not c2["bid"]: zero_legs.append(f"CE({k2}) bid=0")
    if not p2["ask"]: zero_legs.append(f"PE({k2}) ask=0")
    if zero_legs:
        raise ValueError(f"Zero price leg(s) in BOX LONG ({k1}/{k2}): {', '.join(zero_legs)}")

    box_long_price = (
        c1["ask"]      # buy Call K1
        - p1["bid"]    # sell Put K1
        - c2["bid"]    # sell Call K2
        + p2["ask"]    # buy Put K2
    )

    legs = {
        "buy_call_k1": {
            "contract": c1["display_name"],
            "price_used": c1["ask"]
        },
        "sell_put_k1": {
            "contract": p1["display_name"],
            "price_used": p1["bid"]
        },
        "sell_call_k2": {
            "contract": c2["display_name"],
            "price_used": c2["bid"]
        },
        "buy_put_k2": {
            "contract": p2["display_name"],
            "price_used": p2["ask"]
        }
    }

    return box_long_price, legs

File: crates/nse_pricing/src/box_pricer.rs
//! Direct port of `box_pricer.py`. Inputs are rupee floats (best bid/ask,
//! already divided down from the wire's paise-scaled integers), matching
//! this codebase's `nse_fo.py`-derived convention -- see `bid_ask()` in
//! `pricing_fn.rs` for where that conversion happens.

/// Box sell (short box): `-Call(K1) + Put(K1) + Call(K2) - Put(K2)`.
///
/// Execution logic, same as the Python docstring:
/// - `-Call(K1)` -- sell Call(K1) -> use bid
/// - `+Put(K1)`  -- buy Put(K1)   -> use ask
/// - `+Call(K2)` -- buy Call(K2)  -> use ask
/// - `-Put(K2)`  -- sell Put(K2)  -> use bid
///
/// `None` if any required leg price is zero -- `box_pricer.py` raises
/// `ValueError` for this; here it's a skip instead, matching the scheduler's
/// existing "can't price this tick, try again next update" tolerance for
/// incomplete legs.
pub fn box_sell_price(c1_bid: f64, p1_ask: f64, c2_ask: f64, p2_bid: f64) -> Option<f64> {
    if c1_bid == 0.0 || p1_ask == 0.0 || c2_ask == 0.0 || p2_bid == 0.0 {
        return None;
    }
    Some(c1_bid - p1_ask - c2_ask + p2_bid)
}

/// Box long (buy box): `+Call(K1) - Put(K1) - Call(K2) + Put(K2)`.
///
/// Executable price: `C(K1)_ask - P(K1)_bid - C(K2)_bid + P(K2)_ask`.
pub fn box_long_price(c1_ask: f64, p1_bid: f64, c2_bid: f64, p2_ask: f64) -> Option<f64> {
    if c1_ask == 0.0 || p1_bid == 0.0 || c2_bid == 0.0 || p2_ask == 0.0 {
        return None;
    }
    Some(c1_ask - p1_bid - c2_bid + p2_ask)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn box_sell_price_matches_the_formula() {
        // -Call(K1) + Put(K1) + Call(K2) - Put(K2), executable as
        // c1_bid - p1_ask - c2_ask + p2_bid
        let price = box_sell_price(120.0, 30.0, 40.0, 60.0).unwrap();
        assert_eq!(price, 120.0 - 30.0 - 40.0 + 60.0);
    }

    #[test]
    fn box_long_price_matches_the_formula() {
        let price = box_long_price(125.0, 28.0, 38.0, 65.0).unwrap();
        assert_eq!(price, 125.0 - 28.0 - 38.0 + 65.0);
    }

    #[test]
    fn zero_leg_price_is_skipped_not_computed() {
        assert_eq!(box_sell_price(0.0, 30.0, 40.0, 60.0), None);
        assert_eq!(box_sell_price(120.0, 0.0, 40.0, 60.0), None);
        assert_eq!(box_long_price(125.0, 28.0, 0.0, 65.0), None);
        assert_eq!(box_long_price(125.0, 28.0, 38.0, 0.0), None);
    }
}
